client/trino-jdbc/src/test/java/io/trino/jdbc/TestPrestoDriverAssumeUser.java

```diff
@@ -0,0 +1,102 @@
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Map;
+import java.util.Properties;
+
+import static com.google.common.io.Resources.getResource;
+import static java.lang.String.format;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertThrows;
+import static org.testng.Assert.assertTrue;
+
+public class TestPrestoDriverAssumeUser
+{
+    private final String USER = "user";
+    private final String PASSWORD = "password";
+
+    private TestingTrinoServer server;
+
+    @BeforeClass
+    public void setup()
+            throws Exception
+    {
+        server = TestingTrinoServer.builder()
+                .setProperties(ImmutableMap.<String, String>builder()
+                        .put("http-server.authentication.type", "password")
+                        .put("http-server.https.enabled", "true")
+                        .put("http-server.https.keystore.path", getResource("localhost.keystore").getPath())
+                        .put("http-server.https.keystore.key", "changeit")
+                        .build())
+                .build();
+
+        server.installPlugin(new PasswordAuthenticatorPlugin());
+        server.getInstance(Key.get(PasswordAuthenticatorManager.class)).loadPasswordAuthenticator();
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void teardown()
+            throws Exception
+    {
+        server.close();
+    }
+
+    @Test
+    public void testInvalidCredentials()
+    {
+        assertThrows(() -> trySelectCurrentUser(ImmutableMap.of()));
+        assertThrows(() -> trySelectCurrentUser(ImmutableMap.of("user", "invalidUser", "password", PASSWORD)));
+        assertThrows(() -> trySelectCurrentUser(ImmutableMap.of("user", USER, "password", "invalidPassword")));
+        assertThrows(() -> trySelectCurrentUser(ImmutableMap.of("user", "invalidUser", "password", PASSWORD, "requestedUser", USER)));
+    }
+
+    @Test
+    public void testQueryUserNotSpecified()
+            throws SQLException
+    {
+        assertEquals(trySelectCurrentUser(ImmutableMap.of("user", USER, "password", PASSWORD)), USER);
+    }
+
+    @Test
+    public void testAssumeUser()
+            throws SQLException
+    {
+        assertEquals(trySelectCurrentUser(ImmutableMap.of("user", USER, "password", PASSWORD, "requestedUser", "differentUser")), "differentUser");
+    }
+
+    private String trySelectCurrentUser(Map<String, String> additionalProperties)
+            throws SQLException
+    {
+        try (Connection connection = createConnection(additionalProperties);
+                Statement statement = connection.createStatement();
+                ResultSet resultSet = statement.executeQuery("SELECT current_user")) {
```

Comment:
IIRC `current_user` should return authentication user, while `session_user` (which is missing) should return authorization user.

CC: @martint

Reply:
See https://github.com/trinodb/trino/pull/2671#issuecomment-677905852

Reply:
> currently, the "Session#user" on the server seems to be effectively a function of (<authentication header>, <"X-Trino-User" header>), where the later overrides if present. 

@phd3 not exactly, since `X-Trino-User` header is mandatory.

> Once we have #2671, we should have a test around how the "current_user" changes with the new header.

#2671 will impact session state. With session state being stored in all-custom headers, this would mean it's not going to be supported by the clients until they update, unless we implement this as a `Cookie`, wherein, it will end up being independent from the change here.

Anyway, let's move this discussion to that PR, as it's more applicable there, per https://github.com/trinodb/trino/pull/6549#issuecomment-757375379